File: statusline.rb
# frozen_string_literal: true

# Rich statusLine renderer -- the port of statusline-command.rb into ccpool, with two
# changes: the pace bar is now COLOURED (over-pace cells go red -- the monochrome hatch
# was too subtle to read), and it adds the pool $-value. Groups by timescale:
#   now (context window + cache-TTL) · ses (5h) · wk (weekly meter + $ + day-share)
# ANSI is officially supported in statuslines (code.claude.com/docs/en/statusline).
require "json"
require "time"
require_relative "pool"
require_relative "calibration"

module Statusline
  # Honour the NO_COLOR contract (present + non-empty, per no-color.org) and TERM=dumb. We CAN'T gate on tty here: Claude
  # Code invokes the statusLine with a non-tty stdout yet renders ANSI, so tty-gating would strip
  # colour in the primary use. NO_COLOR is the user's explicit opt-out; colour off -> every escape
  # below collapses to "", so the line degrades to plain text (the bar/meter still read via glyphs).
  COLOR = ENV["NO_COLOR"].to_s.empty? && ENV["TERM"] != "dumb"
  def self.ansi(code) = COLOR ? code : ""

  EIGHTHS = [" ", "▏", "▎", "▍", "▌", "▋", "▊", "▉", "█"].freeze
  SOLID = "█"
  TRACK = "░"
  RESET = ansi("\e[0m")
  DIM   = ansi("\e[2m")
  YELLOW = ansi("\e[93m")
  RED   = ansi("\e[91m") # bright, legible on dark backgrounds
  # 24-bit truecolour teal-cyan: theme-INDEPENDENT. A 16-colour code (\e[96m) gets remapped by
  # the terminal palette -- Ghostty rendered cyan as pink -- so we pin the actual RGB. Calm/cool
  # when healthy; the over-pace tail overrides to red. Override via CCPOOL_BAR_COLOR.
  BAR   = ansi(ENV["CCPOOL_BAR_COLOR"] || "\e[38;2;86;182;194m")
  BOLD  = ansi("\e[1m")
  SEP   = " #{DIM}·#{RESET} ".freeze # interpolated -> not auto-frozen by the magic comment
  WEEK  = 7 * 86_400
  CACHE_TTL  = 3600 # statusline staleness display tiers (secs): fresh past this
  CACHE_WARN = 900  # ...dim/warn past this
  CACHE_CRIT = 180  # ...and flag as critically stale past this
  TIER = ENV["USAGE_TIER"] || "max_20x"
  LOG  = File.expand_path(ENV["CCPOOL_STATUSLINE_LOG"] || "~/.claude/statusline.log")

  module_function

  # Best-effort anomaly log, capped to the last 200 lines. The happy path writes NOTHING --
  # only a Claude Code payload-schema change or an unexpected error lands here, so a silent
  # segment drop (below) still leaves a trail: `tail -f ~/.claude/statusline.log`. Never raises.
  def log(level, msg)
    msg  = msg.to_s.gsub(/\s*\n\s*/, " ").slice(0, 500) # one entry is always one line
    prev = File.exist?(LOG) ? File.readlines(LOG) : []
    prev << "#{Time.now.strftime('%F %T')} [#{level}] #{msg}\n"
    File.write(LOG, prev.last(200).join)
  rescue StandardError
    nil
  end

  # True if hash[key] is the expected type. Present-but-wrong-type LOGS a warning (the signal a
  # CC schema change silently dropped a segment); a missing key is silent + expected. Callers
  # && these so a false result skips the segment without touching the wrong-typed value.
  # label defaults to the key; pass a dotted path for nested fields.
  def typed?(hash, key, type, label = key)
    v = hash[key]
    return true if v.is_a?(type)

    log("warn", "#{label} is #{v.class}, expected #{type}") unless v.nil?
    false
  end

  def sev(text, pct, warn:, crit:)
    return "#{RED}#{text}#{RESET}" if pct >= crit
    return "#{YELLOW}#{text}#{RESET}" if pct >= warn

    text
  end

  def fmt_dur(secs)
    secs = 0 if secs.negative?
    d, r = secs.divmod(86_400)
    h, r = r.divmod(3_600)
    m, = r.divmod(60)
    return "#{d}d#{h}h" if d.positive?
    return "#{h}h#{m}m" if h.positive?

    "#{m}m"
  end

  def fmt_size(n)
    return nil unless n.is_a?(Numeric) && n.positive?

    n >= 1_000_000 ? "#{(n / 1_000_000.0).round}M" : "#{(n / 1000.0).round}k"
  end

  # $-left readout, shared by both weekly renderers: "$1.2k" past a grand, else "$47".
  def fmt_dollars(n) = n >= 1000 ? "$#{(n / 1000).round(1)}k" : "$#{n.round}"

  # COLOURED meter: on-pace used cells solid (plain), over-pace used cells RED, partial
  # leading edge in eighths, remaining dim. The red tail IS the "burning too fast" signal.
  def meter(used_frac, pace_frac, width)
    used_w = used_frac * width
    pace_w = (pace_frac || 1.0) * width
    (0...width).map do |i|
      if i + 1 <= used_w
        col = (i + 0.5) >= pace_w ? RED : BAR # over-pace tail red, on-pace fill cyan
        "#{col}#{SOLID}#{RESET}"
      elsif i < used_w
        col = (i + 0.5) >= pace_w ? RED : BAR # partial edge follows the same over-pace rule (no cyan sliver after the red tail)
        "#{col}#{EIGHTHS[[((used_w - i) * 8).round, 1].max]}#{RESET}"
      else
        "#{DIM}#{TRACK}#{RESET}" # dim remaining -> contrast with the fill
      end
    end.join
  end

  # Last-activity epoch + live prompt-cache TTL, detected from the transcript tail's
  # cache buckets (1h on subscription, 5m on paid credits). => [epoch, ttl|nil] or nil.
  def cache_state(path)
    return nil unless path.is_a?(String) && File.exist?(path)

    tail = File.open(path, "rb") do |f|
      f.seek([f.size - 32_768, 0].max)
      f.read
    end
    lines = tail.to_s.lines
    lines.shift if tail.bytesize >= 32_768 && lines.size > 1
    entries = lines.filter_map { |l| JSON.parse(l) rescue nil }.grep(Hash)

    tse = entries.rfind { |e| e["timestamp"].is_a?(String) } or return nil
    ts = (Time.parse(tse["timestamp"]).to_i rescue return nil) # rubocop:disable Style/RedundantParentheses -- parens are REQUIRED to scope `rescue return`
    ttl = nil
    entries.reverse_each do |e|
      cc = e.dig("message", "usage", "cache_creation")
      next unless cc.is_a?(Hash)

      if cc["ephemeral_5m_input_tokens"].to_i.positive?
        ttl = 300
        break
      elsif cc["ephemeral_1h_input_tokens"].to_i.positive?
        ttl = 3_600
        break
      end
    end
    [ts, ttl]
  rescue StandardError
    nil
  end

  # Render the whole line from the fresh CC payload (rate_limits is account-global, so the
  # payload IS current). `dpp` = cached $/1% (fast; never spawn ccusage in the statusline).
  # rubocop:disable Metrics/AbcSize -- inline segment-assembler (ctx/cache/5h/weekly); linear display code, deliberately not fragmented into helpers
  def render(data, now = Time.now.to_i)
    cols = (ENV["COLUMNS"] || "120").to_i
    rl = typed?(data, "rate_limits", Hash) ? data["rate_limits"] : {}
    dpp = Calibration.read_cache&.dig("dpp")
    now_grp = []
    ses_grp = []
    wk_grp = []

    # context window %
    if typed?(data, "context_window", Hash)
      cw = data["context_window"]
      ctx = typed?(cw, "used_percentage", Numeric, "context_window.used_percentage") ? cw["used_percentage"].to_f : nil
      if ctx
        seg = "ctx #{sev("#{ctx.round}%", ctx.round, warn: 70, crit: 90)}"
        (s = fmt_size(cw["context_window_size"])) && (seg += " #{s}")
        now_grp << seg
      end
    end

    # prompt-cache countdown (only when near expiry)
    if (st = cache_state(data["transcript_path"]))
      ts, ttl = st
      left = ts + (ttl || CACHE_TTL) - now
      if left <= 0
        now_grp << "cache #{BOLD}#{RED}cold#{RESET}"
      elsif left < CACHE_WARN
        col = left < CACHE_CRIT ? "#{BOLD}#{RED}" : YELLOW
        now_grp << "cache #{col}#{fmt_dur(left)}#{RESET}"
      end
    end

    # 5h session
    fh = rl["five_hour"]
    if typed?(rl, "five_hour", Hash) && typed?(fh, "used_percentage", Numeric, "five_hour.used_percentage")
      s = fh["used_percentage"].round
      seg = "ses #{sev("#{s}%", s, warn: 80, crit: 92)}"
      seg += " #{fmt_dur(fh['resets_at'] - now)}" if typed?(fh, "resets_at", Numeric, "five_hour.resets_at")
      ses_grp << seg
    end

    # weekly meter + $ + day-share
    sd = rl["seven_day"]
    if typed?(rl, "seven_day", Hash) && typed?(sd, "used_percentage", Numeric, "seven_day.used_percentage")
      used = sd["used_percentage"].to_f
      width = (cols - 82).clamp(14, 40)
      wknum = sev("#{used.round}%", used.round, warn: 75, crit: 90)
      left = (100 - used) * (dpp || 0)
      dollars = dpp ? " #{DIM}#{fmt_dollars(left)}#{RESET}" : ""
      if typed?(sd, "resets_at", Numeric, "seven_day.resets_at")
        reset = sd["resets_at"]
        pace = Profile.elapsed_fraction(reset - WEEK, now, reset) # same weighting as Pool.pace -> bar agrees with verdict
        days_left = [(reset - now).to_f / 86_400, 0.0001].max
        day = [100 - used, (100 - used) / days_left].min.clamp(0, 100) # burnable %/remaining-day
        wk_grp << "wk #{meter(used / 100.0, pace, width)} #{wknum}#{dollars} #{fmt_dur(reset - now)} #{DIM}day #{day.round}%#{RESET}"
      else
        wk_grp << "wk #{meter(used / 100.0, nil, width)} #{wknum}#{dollars}"
      end
    end

    [now_grp, ses_grp, wk_grp].reject(&:empty?).map { |g| g.join("  ") }.join(SEP)
  end
  # rubocop:enable Metrics/AbcSize

  # Compact one-segment render for EMBEDDING inside another statusline (e.g. as a ccstatusline
  # custom-command widget, which forwards the full payload incl. rate_limits -- verified). Shows
  # ONLY ccpool's differentiator -- pool $-left + pace -- and leaves ctx/5h/model/git to the host,
  # so we don't duplicate what it already renders. "" when there's no weekly window to speak to.
  def render_compact(data, now = Time.now.to_i)
    rl = typed?(data, "rate_limits", Hash) ? data["rate_limits"] : {}
    sd = rl["seven_day"]
    return "" unless typed?(rl, "seven_day", Hash) && typed?(sd, "used_percentage", Numeric, "seven_day.used_percentage")

    used = sd["used_percentage"].to_f
    parts = ["pool #{sev("#{used.round}%", used.round, warn: 75, crit: 90)}"]

    if (dpp = Calibration.read_cache&.dig("dpp")) # cache-only: NEVER spawn ccusage in a render path
      left = (100 - used) * dpp
      parts << "#{DIM}#{fmt_dollars(left)}#{RESET}"
    end

    # pace: over-pace (burning fast) is the red risk signal, under-pace is banked headroom (cyan) --
    # same convention as the meter's red tail, so the compact line agrees with the full one.
    if typed?(sd, "resets_at", Numeric, "seven_day.resets_at")
      reset = sd["resets_at"]
      d = (used - (Profile.elapsed_fraction(reset - WEEK, now, reset) * 100)).round
      parts << (d.positive? ? "#{RED}+#{d}↑#{RESET}" : "#{BAR}#{d}↓#{RESET}") if d.abs >= 1
    end

    parts.join(" ")
  end
end

File: internal/statusline/statusline.go
// Package statusline renders the Claude Code statusLine from a fresh CC payload. The rate_limits %
// is account-global, so the payload IS current. No $ here (see Render); `status` has it. Output
// must stay byte-identical to the committed goldens (conformance/golden/, ANSI included).
//
// Groups by timescale:  now (context window + cache-TTL) · 5h · wk (7 day cells + % + reset).
// ANSI is officially supported in statuslines (code.claude.com/docs/en/statusline).
package statusline

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/SeanLF/ccpool/internal/env"
	"github.com/SeanLF/ccpool/internal/fmtx"
	"github.com/SeanLF/ccpool/internal/profile"
	"github.com/SeanLF/ccpool/internal/rb"
	"github.com/muesli/termenv"
)

const (
	week      = 7 * 86400
	cacheWarn = 900 // prompt-cache seconds left: yellow (not dim) under this
	cacheCrit = 180 // ...and bold red under this
)

// heights are the partial-height cell glyphs (index 0..8, empty to full).
var heights = [...]string{" ", "▁", "▂", "▃", "▄", "▅", "▆", "▇", "█"}

// future is an empty day cell: dim, and a dot rather than a grey background, which reads as a heavy
// block on light terminal themes (and a non-tty statusline can't detect the theme).
const future = "·"

// palette holds the ANSI escapes, each already gated on the colour decision (empty when colour is
// off). Resolved per render so per-call NO_COLOR/TERM/CCPOOL_BAR_COLOR env is honoured.
type palette struct {
	reset, dim, yellow, red, bar, bold, sep string
}

func loadPalette() palette {
	prof := colorProfile()

	// seq is a foreground colour as an SGR prefix, downgraded to the active profile (e.g. 24-bit
	// "38;2;.." -> 256-colour "38;5;.." -> 16-colour "36" -> "" for Ascii). termenv does the matching.
	seq := func(spec string) string {
		s := prof.Color(spec).Sequence(false)
		if s == "" {
			return ""
		}
		return "\x1b[" + s + "m"
	}
	// attr gates a non-colour SGR attribute (reset/dim/bold) on colour being on at all.
	attr := func(code string) string {
		if prof == termenv.Ascii {
			return ""
		}
		return code
	}

	// The bar is dim like the rest of the line while on pace, so red past pace is the only colour
	// that means "look". CCPOOL_BAR_COLOR gives an explicit raw-escape override; either way it is
	// suppressed when colour is off (attr returns "" under Ascii).
	bar := os.Getenv("CCPOOL_BAR_COLOR")
	if bar == "" {
		bar = "\x1b[2m"
	}
	bar = attr(bar)

	p := palette{
		reset:  attr("\x1b[0m"),
		dim:    attr("\x1b[2m"),
		yellow: seq("11"), // bright yellow; already 16-colour, so unchanged across the colour tiers
		red:    seq("9"),  // bright red
		bar:    bar,
		bold:   attr("\x1b[1m"),
	}
	p.sep = " " + p.dim + "·" + p.reset + " "
	return p
}

// colorProfile resolves how much colour to emit. Claude Code invokes the statusLine with a NON-TTY
// stdout yet renders ANSI, so termenv's own auto-detection would strip everything; we force TrueColor
// by default and let CCPOOL_COLOR opt down a tier (a non-tty hook can't auto-detect the real
// terminal). NO_COLOR / TERM=dumb still win: forcing the profile bypasses termenv's NO_COLOR check,
// so it is gated manually. An unknown CCPOOL_COLOR fails open to TrueColor.
func colorProfile() termenv.Profile {
	if noColorEnv() {
		return termenv.Ascii
	}
	switch strings.ToLower(strings.TrimSpace(env.String("CCPOOL_COLOR", ""))) {
	case "256", "8bit":
		return termenv.ANSI256
	case "16", "ansi":
		return termenv.ANSI
	case "ascii", "none", "off":
		return termenv.Ascii
	case "auto":
		return termenv.NewOutput(os.Stdout).Profile // termenv's own detection (strips on a non-tty pipe)
	default: // "", "truecolor", "24bit", or anything unrecognised
		return termenv.TrueColor
	}
}

// noColorEnv preserves the pre-termenv contract exactly: NO_COLOR present AND non-empty, or TERM=dumb.
// (termenv.EnvNoColor also honours CLICOLOR and treats an empty NO_COLOR as set, which would differ.)
func noColorEnv() bool {
	v, ok := os.LookupEnv("NO_COLOR")
	return (ok && v != "") || os.Getenv("TERM") == "dumb"
}

// Render builds the whole line from the CC payload. now is unix seconds. No $ (here or in
// RenderCompact): a glance should answer "am I ahead of pace", and cold reads took "$1.4k" for money
// spent; `status` has it.
func Render(data map[string]any, now int64) string {
	pal := loadPalette()
	prof := profile.Load()

	rl := typedHash(data, "rate_limits", "rate_limits")
	if rl == nil {
		rl = map[string]any{}
	}
	var nowGrp, sesGrp, wkGrp []string

	// context window %
	if cw := typedHash(data, "context_window", "context_window"); cw != nil {
		if ctx, ok := typedNum(cw, "used_percentage", "context_window.used_percentage"); ok {
			r := rb.RoundToInt(ctx)
			seg := pal.quiet("ctx") + " " + sev(pct(r), r, 70, 90, pal)
			if s := fmtSize(cw["context_window_size"]); s != "" {
				seg += " " + pal.dim + s + pal.reset
			}
			nowGrp = append(nowGrp, seg)
		}
	}

	// prompt-cache countdown from CC's own prompt_cache (v2.1.251+): always shown so the line doesn't
	// shift, dim until it's near expiry, where colour becomes the alert. CC re-renders at expires_at,
	// so the flip to cold lands on time without a refresh tick.
	if pc := typedHash(data, "prompt_cache", "prompt_cache"); pc != nil {
		if seen, _ := pc["caching_observed"].(bool); seen {
			warm, _ := pc["warm"].(bool)
			exp, hasExp := typedNum(pc, "expires_at", "prompt_cache.expires_at")
			left := int64(exp) - now
			switch {
			case warm && !hasExp:
				// warm but no expiry to count down: say nothing rather than a false "cold"
			case !warm || left <= 0:
				nowGrp = append(nowGrp, pal.quiet("cache")+" "+pal.bold+pal.red+"cold"+pal.reset)
			case left < cacheWarn:
				col := pal.yellow
				if left < cacheCrit {
					col = pal.bold + pal.red
				}
				nowGrp = append(nowGrp, pal.quiet("cache")+" "+col+cacheLeft(left)+pal.reset)
			default:
				nowGrp = append(nowGrp, pal.quiet("cache "+cacheLeft(left)))
			}
		}
	}

	// 5h session
	if fh := typedHash(rl, "five_hour", "five_hour"); fh != nil {
		if used, ok := typedNum(fh, "used_percentage", "five_hour.used_percentage"); ok {
			s := rb.RoundToInt(used)
			seg := pal.quiet("5h-ses") + " " + sev(pct(s), s, 80, 92, pal)
			if reset, ok := typedNum(fh, "resets_at", "five_hour.resets_at"); ok {
				seg += " " + pal.quiet(resetIn(int64(reset)-now))
			}
			sesGrp = append(sesGrp, seg)
		}
	}

	// weekly: 7 day cells + % + reset
	if sd := typedHash(rl, "seven_day", "seven_day"); sd != nil {
		used, hasUsed := typedNum(sd, "used_percentage", "seven_day.used_percentage")
		resetF, hasReset := typedNum(sd, "resets_at", "seven_day.resets_at")
		// Past-reset guard: a stale payload whose window already reset must not show its old % (mirrors
		// pool.GetWindow dropping reset<=now); suppress the whole weekly segment until the payload catches up.
		if hasUsed && !(hasReset && int64(resetF) <= now) {
			// No %-threshold colour here: pace is the weekly alarm, so red means one thing (97% used an
			// hour before reset is fine). The +N↑ repeats the red cells in text, so the signal survives
			// colour-blindness and NO_COLOR.
			// The label, % and on-pace cells stay quiet; only ahead-of-pace lights up.
			wknum := pct(rb.RoundToInt(used))
			if hasReset {
				reset := int64(resetF)
				pace := prof.ElapsedFraction(reset-week, now, reset)
				d := paceDelta(used, pace)
				seg := pal.quiet("wk") + " " + days(used/100.0, pace, d >= 1, pal) + " "
				if d >= 1 {
					seg += wknum + " " + pal.red + fmt.Sprintf("+%d↑", d) + pal.reset
				} else {
					seg += pal.quiet(wknum)
				}
				wkGrp = append(wkGrp, seg+" "+pal.quiet(resetIn(reset-now)))
			} else {
				wkGrp = append(wkGrp, pal.quiet("wk")+" "+days(used/100.0, 1.0, false, pal)+" "+pal.quiet(wknum))
			}
		}
	}

	return joinGroups(pal, nowGrp, sesGrp, wkGrp)
}

// RenderCompact is the one-segment render for embedding in another statusline: ONLY ccpool's
// differentiator (weekly % + pace), leaving ctx/5h/model/git to the host. "" when there's no weekly
// window to speak to.
func RenderCompact(data map[string]any, now int64) string {
	pal := loadPalette()
	prof := profile.Load()

	rl := typedHash(data, "rate_limits", "rate_limits")
	if rl == nil {
		rl = map[string]any{}
	}
	sd := typedHash(rl, "seven_day", "seven_day")
	if sd == nil {
		return ""
	}
	used, ok := typedNum(sd, "used_percentage", "seven_day.used_percentage")
	if !ok {
		return ""
	}
	// Past-reset guard: a stale payload whose window already reset shows no weekly (mirrors pool).
	if resetF, ok := typedNum(sd, "resets_at", "seven_day.resets_at"); ok && int64(resetF) <= now {
		return ""
	}

	r := rb.RoundToInt(used)
	parts := []string{fmt.Sprintf("pool %d%%", r)} // uncoloured: the pace arrow is the alarm, as in Render

	// pace: over-pace (burning fast) is the red risk signal, under-pace is banked headroom (dim).
	if resetF, ok := typedNum(sd, "resets_at", "seven_day.resets_at"); ok {
		reset := int64(resetF)
		d := paceDelta(used, prof.ElapsedFraction(reset-week, now, reset))
		if d >= 1 || d <= -1 {
			if d > 0 {
				parts = append(parts, fmt.Sprintf("%s+%d↑%s", pal.red, d, pal.reset))
			} else {
				parts = append(parts, fmt.Sprintf("%s%d↓%s", pal.bar, d, pal.reset))
			}
		}
	}

	return strings.Join(parts, " ")
}

// --- rendering helpers ---

// sev is a value's severity: dim while healthy, yellow near its limit, red at it. The line is
// ambient by default, so anything not dim means "look at me".
func sev(text string, pct int, warn, crit int, pal palette) string {
	if pct >= crit {
		return pal.red + text + pal.reset
	}
	if pct >= warn {
		return pal.yellow + text + pal.reset
	}
	return pal.quiet(text)
}

// quiet dims text that should sit in the background: labels, countdowns, healthy values.
func (p palette) quiet(text string) string { return p.dim + text + p.reset }

// Fixed-width fields, like tabular figures, so a value changing width (9% -> 10%, 58m -> 9m)
// doesn't shift every segment after it. Durations zero-pad their inner unit for the same reason,
// which keeps widths steady without visible double spaces.
func pct(n int) string { return fmt.Sprintf("%2d%%", n) }

// resetIn is "↻" plus exactly 5 runes for any real window (up to 7d): "0h15m", "9h59m", then the
// day form from 10h ("0d23h", "6d23h"), since "23h59m" would be 6. Minutes only matter under 10h,
// which is the whole 5h window.
func resetIn(secs int64) string {
	secs = max(secs, 0)
	switch {
	case secs < 36000:
		return fmt.Sprintf("↻%dh%02dm", secs/3600, secs%3600/60)
	default:
		return fmt.Sprintf("↻%dd%02dh", secs/86400, secs%86400/3600)
	}
}

// cacheLeft caps at 59m: the TTL is at most an hour, so only the first second after a write would
// read "1h00m" and shift everything after it by two columns.
func cacheLeft(secs int64) string { return fmt.Sprintf("%02dm left", min(max(secs, 0), 3599)/60) }

// fmtSize renders a token count as "1M"/"200k"; "" unless the value is a positive JSON number.
func fmtSize(v any) string {
	if f, ok := rb.Num(v); ok {
		return fmtx.Size(f)
	}
	return ""
}

// paceDelta is used% minus the pace-profile elapsed%, rounded: the same number pool.GetPace (status,
// warn) reports, so every surface agrees on "ahead" (>= 1).
func paceDelta(used, paceFrac float64) int {
	return rb.RoundToInt(used - paceFrac*100)
}

// days is the week as 7 cells, each a seventh of it, filled in order by cumulative use. When ahead
// of pace (paceDelta >= 1), the cells holding use past the pace point are red.
// At most one colour per cell plus empty space, so a cell never needs a background colour.
func days(usedFrac, paceFrac float64, ahead bool, pal palette) string {
	var b strings.Builder
	for k := range 7 {
		f := min(max(usedFrac*7-float64(k), 0), 1)
		n := rb.RoundToInt(f * 8)
		if n == 0 && f > 1e-9 {
			n = 1 // any use shows, however small
		}
		if n == 0 {
			b.WriteString(pal.dim + future + pal.reset)
			continue
		}
		col := pal.bar
		if ahead && float64(k+1)/7 > paceFrac {
			col = pal.red
		}
		b.WriteString(col + heights[n] + pal.reset)
	}
	return b.String()
}

// joinGroups joins every segment with SEP. The groups (now, 5h, week) only order the segments; a
// wider gap inside a group read as an uneven space rather than a grouping.
func joinGroups(pal palette, groups ...[]string) string {
	var all []string
	for _, g := range groups {
		all = append(all, g...)
	}
	return strings.Join(all, pal.sep)
}

// --- typed payload access (mirrors statusline.rb typed?) ---

// typedHash returns m[key] as an object, or nil if absent/null/wrong-typed. A present-but-wrong
// value logs an anomaly (the signal a CC schema change silently dropped a segment).
func typedHash(m map[string]any, key, label string) map[string]any {
	v, present := m[key]
	if v == nil {
		return nil // absent or JSON null: silent + expected
	}
	if h, ok := v.(map[string]any); ok {
		return h
	}
	if present {
		diag.Warn("segment not an object", "field", label, "got", fmt.Sprintf("%T", v))
	}
	return nil
}

// typedNum returns m[key] as a float64 if it is a JSON number, else (0,false). Wrong-typed logs.
func typedNum(m map[string]any, key, label string) (float64, bool) {
	v, present := m[key]
	if v == nil {
		return 0, false
	}
	if n, ok := v.(json.Number); ok {
		if f, err := n.Float64(); err == nil {
			return f, true
		}
	}
	if present {
		diag.Warn("segment not a number", "field", label, "got", fmt.Sprintf("%T", v))
	}
	return 0, false
}
